src/validation/tests/assignment_validation_tests.rs

```diff
@@ -962,3 +962,100 @@ fn string_type_alias_assignment_can_be_validated() {
 
     assert_snapshot!(diagnostics);
 }
+
+#[test]
+fn integral_promotion_in_expression_does_not_cause_downcast_warning() {
+    let diagnostics = parse_and_validate_buffered(
+        "
+        FUNCTION main : INT
+        VAR
+            a, b, c : SINT;
+        END_VAR
+
+            a := a + b + c + 100;
+        END_FUNCTION
+        ",
+    );
+
+    assert!(diagnostics.is_empty())
+}
+
+#[test]
+fn downcast_will_report_bigger_types_in_expression() {
+    let diagnostics = parse_and_validate_buffered(
+        "
+        FUNCTION main : INT
+        VAR
+            a, b, c : SINT;
+            d, e : DINT;
+        END_VAR
+            a := a + b + c + d + e;
+        END_FUNCTION
+        ",
+    );
+
+    assert_snapshot!(diagnostics)
+}
+
+#[test]
+fn literals_out_of_range_for_lhs_will_result_in_downcast_warning() {
+    let diagnostics = parse_and_validate_buffered(
+        "
+        FUNCTION main : INT
+        VAR
+            a : SINT;
+            b : USINT;
+        END_VAR
+
+            a := a + 255 + 1000;    // literals out of range of i8 -> warning
+            b := b + 255;           // 255 is in range of u8 -> no warning, but will silently overflow for anything other than b == 0. Not sure if there is a better way to guard against this
+            a := a + b;             // B is same size as a, but unsigned -> warning
+            a := a + USINT#100;     // will fit into a, but is cast to unsigned type -> warning
+        END_FUNCTION
+        ",
+    );
+
+    assert_snapshot!(diagnostics)
+}
+
+
+#[test]
+fn literals_out_of_range_inside_unary_expressions_will_cause_no_warning() {
+    // GIVEN literals behind unary expressions (which will be annotated as DINT)
+    // WHEN we validate
+    let diagnostics = parse_and_validate_buffered(
+        "
+        FUNCTION main : INT
+        VAR
+            a : INT;
+        END_VAR
+            a := a;
+            a := a + 254;
+            a := a + 254 - 20;
+            a := a + 254 + (-(10+10));  //rewrite this as a unary expression
+        END_FUNCTION
+        ",
+    );
+    //WE EXPECT NO VALIDATION PROBLEMS
+    assert_snapshot!(diagnostics)
+}
+
+#[test]
+fn literals_out_of_range_in_a_modulo_operation_cannot_exceed_the_left_operand() {
+    // GIVEN an expression INT MOD DINT
+    // WHEN we validate
+    let diagnostics = parse_and_validate_buffered(
+        "
+        FUNCTION main : INT
+        VAR
+            a : INT;
+            d : DINT := 9;
+        END_VAR
+            a := a + (a MOD d);  // I know this is bullshit - lets ignore it
```

Comment:
shall we handle this correctly or is this silly? This is probably wrongfully annotated. I'm open for discussions.

Reply:
Turns out this is problematic with `SEL` and `MUX` (and most likely all builtins) aswell. I'll have to dig a bit deeper on these cases.